lib/auth/transport_credentials.go

```diff
@@ -0,0 +1,243 @@
+		return trace.BadParameter("parameter TransportCredentials required")
+	case c.TransportCredentials.Info().SecurityProtocol != "tls":
+		return trace.BadParameter("the TransportCredentials must be a tls security protocol, got %s", c.TransportCredentials.Info().SecurityProtocol)
+	case c.UserGetter == nil:
+		return trace.BadParameter("parameter UserGetter required")
+	case c.Authorizer != nil && c.UserGetter == nil:
+		return trace.BadParameter("a UserGetter is required to use validate identities with an Authorizer")
+	case c.Enforcer != nil && c.Authorizer == nil:
+		return trace.BadParameter("both a UserGetter and an Authorizer are required to enforce connection limits with an Enforcer")
+	default:
+		return nil
+	}
+}
+
+// TransportCredentials is a [credentials.TransportCredentials] that
+// enforces mTLS and retrieves the [IdentityGetter] for use by middleware
+// to perform authorization.
+type TransportCredentials struct {
+	credentials.TransportCredentials
+
+	userGetter UserGetter
+	authorizer Authorizer
+	enforcer   ConnectionEnforcer
+}
+
+// NewTransportCredentials returns a new TransportCredentials
+func NewTransportCredentials(cfg TransportCredentialsConfig) (*TransportCredentials, error) {
+	if err := cfg.Check(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &TransportCredentials{
+		TransportCredentials: cfg.TransportCredentials,
+		userGetter:           cfg.UserGetter,
+		authorizer:           cfg.Authorizer,
+		enforcer:             cfg.Enforcer,
+	}, nil
+}
+
+// IdentityInfo contains the auth information and identity
+// for an authenticated TLS connection. It implements the
+// [credentials.AuthInfo] interface and is returned from
+// [TransportCredentials.ServerHandshake].
+type IdentityInfo struct {
+	// TLSInfo contains TLS connection information.
+	*credentials.TLSInfo
+	// IdentityGetter provides a mechanism to retrieve the
+	// identity of the client.
+	IdentityGetter IdentityGetter
+	// AuthContext contains information about the traits and roles
+	// that an identity may have. This will be unset if the
+	// [TransportCredentialsConfig.Authorizer] provided to [NewTransportCredentials]
+	// was nil.
+	AuthContext *Context
+}
+
+// ServerHandshake does the authentication handshake for servers. It returns
+// the authenticated connection and the corresponding auth information about
+// the connection.
+// At minimum the TLS handshake is performed and the identity is built from
+// the [tls.ConnectionState]. If the TransportCredentials is configured with
+// and Authorizer and ConnectionEnforcer then additional session controls are
+// applied before the handshake completes.
+func (c *TransportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, tlsInfo, err := c.performTLSHandshake(rawConn)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	identityGetter, err := c.userGetter.GetUser(tlsInfo.State)
+	if err != nil {
+		conn.Close()
```

Comment:
nit: to prevent placing conn.Close call in each error != case we can just 
add a helper function like:
```go
(c *TransportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
   conn, authInfo, err := c.serverHandshake(rawConn) 
   if err != nil {
     conn.Close()
     return nil, nil. trace.Wrap(err)
   }
}
```